core/src/main/scala/akka/kafka/internal/SingleSourceLogic.scala

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.kafka.internal
+
+import akka.actor.{ActorRef, ExtendedActorSystem, Terminated}
+import akka.kafka.KafkaConsumerActor.RequestMessages
+import akka.kafka.Subscription.{Assignment, AssignmentWithOffset, TopicSubscription, TopicSubscriptionPattern}
+import akka.kafka.{ConsumerSettings, KafkaConsumerActor, Subscription}
+import akka.stream.stage.GraphStageLogic.StageActor
+import akka.stream.stage.{GraphStageLogic, OutHandler}
+import akka.stream.{ActorMaterializer, SourceShape}
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.common.TopicPartition
+
+import scala.annotation.tailrec
+
+private[kafka] abstract class SingleSourceLogic[K, V, Msg](
+    val shape: SourceShape[Msg],
+    settings: ConsumerSettings[K, V],
+    subscription: Subscription
+) extends GraphStageLogic(shape) with PromiseControl with MessageBuilder[K, V, Msg] {
+  var consumer: ActorRef = _
+  var self: StageActor = _
+  var tps = Set.empty[TopicPartition]
+  var buffer: Iterator[ConsumerRecord[K, V]] = Iterator.empty
+  var requested = false
+  implicit lazy val ec = ActorMaterializer.downcast(materializer).executionContext
+
+  override def preStart(): Unit = {
+    super.preStart()
+    setKeepGoing(true)
+
+    consumer = {
+      val as = ActorMaterializer.downcast(materializer).system.asInstanceOf[ExtendedActorSystem]
```

Comment:
probably not, in this case I would have probably just broke the long call chain into multiple lines, but a val might be fine, but maybe change its name then to `extendedActorSystem` to add readability (because I need to currently need to read the whole thing anyway to figure out what `as` means so the val buys nothing really).